### ISSUE TITLE
Possibility to use property as long and lat values

### ISSUE THREAD

**Issue by mingan666** [mrgadget/node-red-contrib-eztimer#39]
Is there anyway to use a property as long and values?

I would like to use some global setting for these values instead of having to set them each time a new eztimer is dropped on the canvas...

Would be great if it was possible to use global values for example from home assistant.

**Comment by mrgadget** [comment]
You mean lat/long yeah?  I agree - it's something I've thought of before.  Not sure how to do it globally - will have a look at it when I have some time.

In the meantime, I get around the frustration by copy/pasting the node instead of dropping a new one.

**Comment by mingan666** [comment]
Did a bunch of googling and came over this example of using long/lat from home assistant within node-red:
   [https://community-assets.home-assistant.io/original/3X/5/6/56737125793f7c2bc24db799fc4bbe23380d3296.png](https://community-assets.home-assistant.io/original/3X/5/6/56737125793f7c2bc24db799fc4bbe23380d3296.png)

But I'd be happy with other single time config solutions also!

I'll be using copy/paste for now.

**Comment by mrgadget** [comment]
Ok great, thanks for that.  I use NodeRed nested inside HA too, so that certainly looks like an option.  I envisage this would look like a dropdown just above the lat/long giving options of `Manual` or `HomeAssistant Zone`, and if the HA zone is picked, a second dropdown appears in place of the lat/long, giving the HA zones to pick from. 

My `Home` zone was actually called `home_2` - not sure why exactly, perhaps it's because it an old install that's been upgraded and I've moved from YAML-defined zones to the newer UI-defined zones.

Anyway - your thoughts?

**Comment by mingan666** [comment]
I like that idea!

**Comment by mingan666** [comment]
I also think it should be possible to chose automatically one of the zones starting with home?
So there is a default so to say if there is only one?
If the identifier is changed or if there are more you could still choose the correct one manually from the dropdown.

**Comment by mrgadget** [comment]
Hey - would appreciate if you could give this a shot and report back before I push it out to the wider user-base.

It's not in NPM yet, so you'll need to grab index.js and index.html and replace them in the node on your system.

Thanks!

**Comment by mingan666** [comment]
I'm happy to test that for you, however I'm kinda new to node-red.
Where do you want me to put these files?
Not sure if this helps, but I normally run npm install in /data in my docker container.

**Comment by mrgadget** [comment]
Hard with every install being different - you're looking for a folder called `node_modules/node-red-contrib-eztimer`.  Within that folder you'll find the `index.html` and `index.js` files which you just need to replace with the ones in the code section of this repo.  You'll also need to restart Node-Red afterwards.  Hope that helps!

**Comment by mingan666** [comment]
Looks good so far!
Really like the way it automatically picks position from home assistant.
Kinda hard to test the actual working of the node, but I'll run with it for a while and get back to you if I run into any issues,
Thank you!

**Comment by mrgadget** [comment]
Great - good to hear it works for you. Going to do a small tweak to hide the source selection combo if it can't get anything from HomeAssistant I think - then put an item in the readme outlining how it works. 

Thanks again for pointing me in this direction!

**Comment by JasonSwindle** [comment]
I've tested this as well, and I LOVE IT. One less hardcoded value.

**Comment by mrgadget** [comment]
Haha good to hear.  I will add at this stage it's still hardcoded in a way...  The value is only obtained from HomeAssistant at time of editing the node.  This is something that _could_ change, but more testing is required.

**Comment by JasonSwindle** [comment]
I could see a value in having it change. Having HASS in an RV would be great. 🙂

---
Thanks,
Jason Swindle
________________________________

**Comment by mrgadget** [comment]
Well you're not wrong there. 

I'll see tonight if I can look up the zone x seconds after startup (worried about HA not having populated the global variable containing the info yet), and update the node config. Would then have to re-init the eztimer node too.

**Comment by mingan666** [comment]
I didn't think of those aspects. That might actually be a problems
What happens if the system restarts and the docker containers are started in the wrong order?
I guess that means the node will not work as intended?
Will it resolve itself the next day? I.e. when calculating the next sunset time?

**Comment by mingan666** [comment]
@JasonSwindle Isn't the zone information in home assistant static?

**Comment by mrgadget** [comment]
No it wouldn't cause any issues.  The lat/long values would still be stored in the node config - the update, when I implement it, will just update this value, if the source is available - so if there are any issues obtaining values due to container startup order etc, there will be no operational impact to eztimer nodes.

@mingan666 the zone information is 'static', but there's nothing to stop a user moving a zone in HA, which eztimer, in it's current guise, would never see.

**Comment by mingan666** [comment]
Thanks for the explanation!

**Comment by JasonSwindle** [comment]
> @JasonSwindle Isn't the zone information in home assistant static? 


@mrgadget Does this change anything?
https://www.home-assistant.io/blog/2019/06/26/release-95/#rv-assistant :)

https://www.home-assistant.io/integrations/homeassistant/#service-homeassistantset_location

**Comment by mrgadget** [comment]
@JasonSwindle well... perhaps another option for _refresh frequency_?  I'm still yet to determine if/how I can update the config from within the node - I'm sure it can be done though.

Getting the zones was a little convoluted as the node config stuff is client-side, so I needed to do a JSON call to a webservice running server-side to get the list of zones - updating them should be easier, as that's server-side already.

Hopefully get some time tonight after family duties are done to have a tinker.

**Comment by mrgadget** [comment]
Updating is still a WIP, but published what's done in 1.2.5, will look at the rest when I get time and that'll be the next release.

**Comment by zachowj** [comment]
https://github.com/mrgadget/node-red-contrib-eztimer/blob/ae0397fe411288198a8f52d660e79ac20c3e2bb7/index.js#L61

`homeAssistant` can be undefined here if the users server name is not exactly "Home Assistant`. This also doesn't take into account the possibilities of two different servers.

Maybe something like 

```javascript
var servers = Object.keys(ha);
var multipleServers = servers.length > 1;
for (server in servers) {
  for (element in ha[server].states) {
    var zone = ha[server].states[element];
    if (element.substring(0, 4) == "zone") {
      var z = {
        entity_id: zone.entity_id,
        name: multipleServers
          ? `${server}: ${zone.attributes.friendly_name}`
          : zone.attributes.friendly_name,
        latitude: zone.attributes.latitude,
        longitude: zone.attributes.longitude,
      };
      if (z.entity_id.substring(0, 9) == "zone.home") zones.unshift(z);
      else zones.push(z);
    }
  }
}
```

**Comment by mrgadget** [comment]
Interesting @zachowj - thanks for that.  Didn't contemplate multiple servers - do you think this needs another combo box for selecting the server?

**Comment by zachowj** [comment]
> do you think this needs another combo box for selecting the server?

That looks to be a good way to do it given your current implementation.

**Comment by mrgadget** [comment]
Inserted your code @zachowj and noticed that I don't need that dropdown - thanks to the iteration over the arrays you suggest, and prefixing the zone names - that works great.  Appreciate the code snippit.

Do you have multiple servers? Could you give it a test?  Latest code on here now includes support, and it still works fine for me with my single default-named server.

Ready to push to npm as part of 1.2.7 if we test ok.  Cheers.

**Comment by zachowj** [comment]
Worked fine in my quick test. To test it yourself you can just create a new server config and just name it differently and use the same host and access token.

**Comment by mrgadget** [comment]
Ok cool, looks good, published 1.2.7.  Thanks for the testing!
